assembly-syntax/src/sema/context.rs

```diff
@@ -47,53 +47,136 @@ impl AnalysisContext {
+                },
+                Err(err) => {
+                    self.errors.push(err);
+                },
+            }
+        }
+    }
+
+    fn const_eval(&mut self, value: &ConstantExpr) -> Result<ConstantExpr, SemanticAnalysisError> {
+        /// Represents the type of a continuation to apply during evaluation
+        enum Cont {
+            /// We have reached an anonymous expression to evaluate
+            Eval(ConstantExpr),
+            /// We have finished evaluating the operands of a constant op, and must now apply the
+            /// operation to them, pushing the result on the operand stack.
+            Apply(Span<ConstantOp>),
+            /// We have finished evaluating a reference to another constant and are returning
+            /// its value on the operand stack
+            Return(Ident),
         }
+
+        // The operand stack
+        let mut stack = Vec::with_capacity(8);
+        // The continuation stack
+        let mut continuations = Vec::with_capacity(8);
+        // Start evaluation from the root expression
+        continuations.push(Cont::Eval(value.clone()));
+        // Keep track of the stack of constants being expanded during evaluation
+        //
+        // Any time we reach a reference to another constant that requires evaluation, we check if
+        // we're already in the process of evaluating that constant. If so, then a cycle is present
+        // and we must raise an eval error.
+        let mut evaluating = SmallVec::<[_; 8]>::new_const();
+
+        while let Some(next) = continuations.pop() {
+            match next {
+                Cont::Eval(
+                    expr @ (ConstantExpr::Int(_)
+                    | ConstantExpr::String(_)
+                    | ConstantExpr::Word(_)
+                    | ConstantExpr::Hash(..)),
+                ) => {
+                    stack.push(expr);
+                },
+                Cont::Eval(ConstantExpr::Var(name)) => {
+                    if evaluating.contains(&name) {
+                        return Err(SemanticAnalysisError::ConstEvalCycle {
+                            start: evaluating[0].span(),
+                            detected: name.span(),
+                        });
+                    } else {
+                        evaluating.push(name.clone());
+                    }
+                    continuations.push(Cont::Return(name.clone()));
+                    continuations.push(Cont::Eval(self.get_constant(&name)?.clone()));
+                },
+                Cont::Eval(ConstantExpr::BinaryOp { span, op, lhs, rhs, .. }) => {
+                    continuations.push(Cont::Apply(Span::new(span, op)));
+                    continuations.push(Cont::Eval(*lhs));
+                    continuations.push(Cont::Eval(*rhs));
+                },
+                Cont::Apply(op) => {
+                    let lhs = stack.pop().unwrap().expect_int();
+                    let rhs = stack.pop().unwrap().expect_int();
```

Comment:
Here as well ?

Reply:
See my comment above - basically it doesn't matter as long as they agree, but for the sake of consistency I think it makes sense to expect `rhs` on top.